pkg/usagestats/stats.go

```diff
@@ -0,0 +1,352 @@
+package usagestats
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"expvar"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/loki/pkg/util/build"
+
+	"github.com/cespare/xxhash/v2"
+	jsoniter "github.com/json-iterator/go"
+	prom "github.com/prometheus/prometheus/web/api/v1"
+	"go.uber.org/atomic"
+)
+
+var (
+	httpClient    = http.Client{Timeout: 5 * time.Second}
+	usageStatsURL = "https://stats.grafana.org/loki-usage-report"
+	statsPrefix   = "github.com/grafana/loki/"
+	targetKey     = "target"
+	editionKey    = "edition"
+)
+
+// Report is the JSON object sent to the stats server
+type Report struct {
+	ClusterID              string    `json:"clusterID"`
+	CreatedAt              time.Time `json:"createdAt"`
+	Interval               time.Time `json:"interval"`
+	Target                 string    `json:"target"`
+	prom.PrometheusVersion `json:"version"`
+	Os                     string                 `json:"os"`
+	Arch                   string                 `json:"arch"`
+	Edition                string                 `json:"edition"`
+	Metrics                map[string]interface{} `json:"metrics"`
+}
+
+// sendReport sends the report to the stats server
+func sendReport(ctx context.Context, seed *ClusterSeed, interval time.Time) error {
+	report := buildReport(seed, interval)
+	out, err := jsoniter.MarshalIndent(report, "", " ")
```

Comment:
I thought it's gonna be Prometheus metrics. What's the reason for a custom API and store?

Reply:
It's very hard to read a Prometheus metric. And I needed more stats like counter, min,max, string !